Moving concurrent-search out of sandbox plugin to core behind feature flag

sohami opened the pull request:
### Description
Moving concurrent-search out of sandbox plugin to server module behind a feature flag

### Issues Resolved
https://github.com/opensearch-project/OpenSearch/issues/6859

### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [ ] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

sohami commented:
@reta @nknize Please help with review when you get chance

sohami commented:
The test failures seems unrelated to this PR:

1. **org.opensearch.index.reindex.ReindexBasicTests.classMethod** - Looks like some race in threadpool getting shutdown and task being submitted for `fail_stale_replica`. Similar stack is seen [here](https://github.com/opensearch-project/OpenSearch/issues/7103)
```
 thg 4 18, 2023 3:10:50 SA org.apache.lucene.store.MemorySegmentIndexInputProvider <init>
INFO: Using MemorySegmentIndexInput with Java 19; to disable start with -Dorg.apache.lucene.store.MMapDirectory.enableMemorySegments=false
thg 4 18, 2023 3:11:27 SA com.carrotsearch.randomizedtesting.RandomizedRunner$QueueUncaughtExceptionsHandler uncaughtException
WARNING: Uncaught exception in thread: Thread[#196,opensearch[node_s1][generic][T#6],5,TGRP-ReindexBasicTests]
org.opensearch.core.concurrency.OpenSearchRejectedExecutionException: rejected execution of java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask@1a874b8f[Not completed, task = java.util.concurrent.Executors$RunnableAdapter@1a8a24ec[Wrapped task = [threaded] fail_stale_replica]] on org.opensearch.threadpool.Scheduler$SafeScheduledThreadPoolExecutor@2dd992da[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 191]
	at __randomizedtesting.SeedInfo.seed([1060866F1479C3B8]:0)
	at org.opensearch.common.util.concurrent.OpenSearchAbortPolicy.rejectedExecution(OpenSearchAbortPolicy.java:67)
	at java.base/java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:841)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor.delayedExecute(ScheduledThreadPoolExecutor.java:340)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor.schedule(ScheduledThreadPoolExecutor.java:562)
	at org.opensearch.threadpool.ThreadPool.schedule(ThreadPool.java:424)
	at org.opensearch.common.util.concurrent.AbstractAsyncTask.rescheduleIfNecessary(AbstractAsyncTask.java:109)
	at org.opensearch.common.util.concurrent.AbstractAsyncTask.run(AbstractAsyncTask.java:174)
	at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:747)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1589)

REPRODUCE WITH: ./gradlew ':modules:reindex:test' --tests "org.opensearch.index.reindex.ReindexBasicTests" -Dtests.seed=1060866F1479C3B8 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=en -Dtests.timezone=Etc/UTC -Druntime.java=19
```

2. **org.opensearch.indices.replication.SegmentReplicationRelocationIT.testRelocateWhileContinuouslyIndexingAndWaitingForRefresh** - The search failed as the index seems to be deleted as part of test cleanup. The [second run](https://build.ci.opensearch.org/job/gradle-check/14175/testReport/junit/org.opensearch.indices.replication/SegmentReplicationRelocationIT/testRelocateWhileContinuouslyIndexingAndWaitingForRefresh_2/) was successful

```
Stacktrace
Failed to execute phase [query], all shards failed
	at __randomizedtesting.SeedInfo.seed([1060866F1479C3B8:9C145E582A6F6340]:0)
	at app//org.opensearch.action.search.AbstractSearchAsyncAction.onPhaseFailure(AbstractSearchAsyncAction.java:664)
	at app//org.opensearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:372)
	at app//org.opensearch.action.search.AbstractSearchAsyncAction.onPhaseDone(AbstractSearchAsyncAction.java:699)
	at app//org.opensearch.action.search.AbstractSearchAsyncAction.onShardFailure(AbstractSearchAsyncAction.java:472)
	at app//org.opensearch.action.search.AbstractSearchAsyncAction.lambda$performPhaseOnShard$0(AbstractSearchAsyncAction.java:273)
	at app//org.opensearch.action.search.AbstractSearchAsyncAction$2.doRun(AbstractSearchAsyncAction.java:350)
	at app//org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52)
	at app//org.opensearch.threadpool.TaskAwareRunnable.doRun(TaskAwareRunnable.java:78)
	at app//org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52)
	at app//org.opensearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:59)
	at app//org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:806)
	at app//org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52)
	at java.base@19.0.2/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base@19.0.2/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base@19.0.2/java.lang.Thread.run(Thread.java:1589)

REPRODUCE WITH: ./gradlew ':server:internalClusterTest' --tests "org.opensearch.indices.replication.SegmentReplicationRelocationIT.testRelocateWhileContinuouslyIndexingAndWaitingForRefresh" -Dtests.seed=1060866F1479C3B8 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=sr-Latn-ME -Dtests.timezone=Pacific/Fiji -Druntime.java=19
```

reta commented on server/src/main/java/org/opensearch/search/query/QueryPhase.java:
```diff
@@ -88,7 +89,9 @@ public class QueryPhase {
     // TODO: remove this property
     public static final boolean SYS_PROP_REWRITE_SORT = Booleans.parseBoolean(System.getProperty("opensearch.search.rewrite_sort", "true"));
-    public static final QueryPhaseSearcher DEFAULT_QUERY_PHASE_SEARCHER = new DefaultQueryPhaseSearcher();
+    public static final QueryPhaseSearcher DEFAULT_QUERY_PHASE_SEARCHER = FeatureFlags.isEnabled(FeatureFlags.CONCURRENT_SEGMENT_SEARCH)
```
The `DEFAULT_QUERY_PHASE_SEARCHER` should stay unchanged and should not be impacted by the  `FeatureFlags`

sohami replied:
My thought is eventually `ConcurrentQueryPhaseSearcher` will become default in core and there won't be a need for separate `DefaultQueryPhaseSearcher` hence keeping this assignment behind feature flag

sohami replied:
We are honoring the `QueryPhaseSearcher` if provided via a plugin. The `DEFAULT_QUERY_PHASE_SEARCHER` instance is used only when no `QueryPhaseSearcher` is explicitly provided.

reta replied:
What I am trying to say here is that the feature flag got leaked everywhere and makes it difficult to understand what is going to happen. The decisions should be taken ideally in one place, in the `SearchModule`: check plugins -> check feature flag -> use one or none

sohami replied:
moved to `SearchModule`

reta commented on server/src/main/java/org/opensearch/search/SearchModule.java:
```diff
@@ -347,7 +343,6 @@ public SearchModule(Settings settings, List<SearchPlugin> plugins) {
         registerShapes();
         registerIntervalsSourceProviders();
         queryPhaseSearcher = registerQueryPhaseSearcher(plugins);
-        indexSearcherExecutorProvider = registerIndexSearcherExecutorProvider(plugins);
```
Please revert this change: we exposed the extension point in the `SearchPlugin` and should respect that. However, if no executor is provided by the plugins, we could consult the feature flag and provide the thread pool executor.

reta commented on server/src/main/java/org/opensearch/node/Node.java:
```diff
@@ -1014,8 +1013,7 @@ protected Node(
                 responseCollectorService,
-                circuitBreakerService,
-                searchModule.getIndexSearcherExecutor(threadPool)
+                circuitBreakerService
```
Please revert this change, the `IndexSearcherExecutor` could be provided by the plugins since we opened this extension point.

reta commented:
> @reta @nknize Please help with review when you get chance

Thanks @sohami , left some comments, just to summarize - we should not ignore the plugin extensions which we already provided (and those could be used already) but we could change the defaults depending on the feature flag.

sohami commented:
> > @reta @nknize Please help with review when you get chance
> 
> Thanks @sohami , left some comments, just to summarize - we should not ignore the plugin extensions which we already provided (and those could be used already) but we could change the defaults depending on the feature flag.

@reta Thanks for the review feedback. I removed the executor related extension because that was only used for concurrent search flow. The concurrent search flow was supported as `experimental sandbox` plugin so far, which means the executor probably is not provided by any other usage (as that will be not used). But on thinking again seems like there can be another parallel implementation as a plugin which may be using it.

reta commented:
> But on thinking again seems like there can be another parallel implementation as a plugin which may be using it.

Correct, we also cannot just remove public APIs (like plugins fe), those should go through the deprecation cycle first and than removed in next major release.

sohami commented:
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** FAILURE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/14272/
> * **CommitID:** [91f23c5](https://github.com/opensearch-project/OpenSearch/commit/91f23c52efa4acf62de2660c98c1b4cf7f160e38)
>   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Because of below which seems to be related to new release branch. Have rebased to see if that fixes this

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':distribution:bwc:minor:buildBwcLinuxTar'.
> Building 2.7.0 didn't generate expected file /var/jenkins/workspace/gradle-check/search/distribution/bwc/minor/build/bwc/checkout-2.x/distribution/archives/linux-tar/build/distributions/opensearch-min-2.7.0-SNAPSHOT-linux-x64.tar.gz
```

reta commented on server/src/main/java/org/opensearch/search/SearchService.java:
```diff
@@ -323,7 +324,15 @@ public SearchService(
         );
-        this.indexSearcherExecutor = indexSearcherExecutor;
+        // use the index searcher executor provided by plugin or the default one in core
+        if (indexSearcherExecutor != null) {
```
Please revert this change, it goes back to https://github.com/opensearch-project/OpenSearch/pull/7203/files#r1171283310, the decision what to use should be taken in single place by `SearchModule`: check plugins -> check feature flag -> use one or none.

reta commented on server/src/main/java/org/opensearch/threadpool/ThreadPool.java:
```diff
@@ -256,6 +258,7 @@ public ThreadPool(
         );
         builders.put(Names.TRANSLOG_SYNC, new FixedExecutorBuilder(settings, Names.TRANSLOG_SYNC, allocatedProcessors * 4, 10000));
         builders.put(Names.REMOTE_PURGE, new ScalingExecutorBuilder(Names.REMOTE_PURGE, 1, halfProcMaxAt5, TimeValue.timeValueMinutes(5)));
+        builders.put(Names.INDEX_SEARCHER, new FixedExecutorBuilder(settings, Names.INDEX_SEARCHER, allocatedProcessors, 1000, false));
```
It should be done in `SearchModule` by the `ExecutorServiceProvider` instance in case the fall back to feature flag kicks in.

sohami replied:
Keeping the builder here but returning the provider in SearchModule which will return `INDEX_SEARCHER` threadpool in case when feature flag is set

reta replied:
Please move it to SearchModule because it leaks outside even if the feature is turned off:

```
curl http://localhost:9200/_cat/thread_pool

...
runTask-0 index_searcher      0 0 0
runTask-0 listener            0 0 0
...

```

sohami replied:
Before `SearchModule` is created all the ExecutorBuilders are already constructed in `ThreadPool`, so don't see a way to add new ExecutorBuilder there. If I just return a `ExecutorService` in `SearchModule` then that won't be tracked by `ThreadPool` and also miss the stats around it. I considered adding Feature flag check in `ThreadPool` class before adding the builder in the map but I think we are trying to avoid that.

reta replied:
Oh sorry, indeed the ExecutorBuilders were coming from PluginService, not a SearchPlugin,  adding feature flag check should make it, thanks.

sohami commented:
> 

Unrelated to this PR and looks to be a flaky test tracked [here](https://github.com/opensearch-project/OpenSearch/issues/6761)

reta commented:
@sohami LGTM, needs @nknize sign off to confirm this is a direction we're moving

sohami commented:
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** FAILURE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/14535/
> * **CommitID:** [1c45cf9](https://github.com/opensearch-project/OpenSearch/commit/1c45cf986c8d1e90788120ebf3545edc7b79fb15)
>   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

There was a test failure for default threadpool type. Added the `index_searcher` pool type behind feature flag to fix this as executor is created behind `FeatureFlag` as well

```
java.lang.IllegalArgumentException: no executor service found for [index_searcher]
	at __randomizedtesting.SeedInfo.seed([E9942ED9E459F96C:FE09E5B6CDE61AE6]:0)
	at org.opensearch.threadpool.ThreadPool.executor(ThreadPool.java:403)
	at org.opensearch.threadpool.TestThreadPool.executor(TestThreadPool.java:81)
	at org.opensearch.threadpool.EvilThreadPoolTests.testExecutionErrorOnDefaultThreadPoolTypes(EvilThreadPoolTests.java:73)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:938)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:974)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:988)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
	at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
	at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
	at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
	at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
```

sohami commented:
> @sohami LGTM, needs @nknize sign off to confirm this is a direction we're moving

@nknize Please let me know if this PR looks good to you.

sohami commented:
> > @sohami LGTM, needs @nknize sign off to confirm this is a direction we're moving
> 
> @nknize Please let me know if this PR looks good to you.

@nknize Reminder for this PR

kotwanikunal commented:
@sohami LGTM. Can you please rebase and resolve the conflicts? 
Also tagging @nknize

kotwanikunal commented:
```
REPRODUCE WITH: ./gradlew ':server:internalClusterTest' --tests "org.opensearch.action.admin.cluster.stats.ClusterStatsIT.testNodeRolesWithDataNodeLegacySettings" -Dtests.seed=9C131CE66EBDF984 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=ar-TN -Dtests.timezone=Africa/Kinshasa -Druntime.java=19

org.opensearch.action.admin.cluster.stats.ClusterStatsIT > testNodeRolesWithDataNodeLegacySettings FAILED
    java.lang.IllegalArgumentException: duplicate element: [cluster_manager]
        at __randomizedtesting.SeedInfo.seed([9C131CE66EBDF984:B56ED209C37D3942]:0)
        at java.****/java.util.ImmutableCollections$Set12.<init>(ImmutableCollections.java:791)
        at java.****/java.util.Set.of(Set.java:487)
        at org.opensearch.action.admin.cluster.stats.ClusterStatsIT.testNodeRolesWithDataNodeLegacySettings(ClusterStatsIT.java:415)

```

Strange! Seems like #7327 again.
@sohami Did you rebase with the latest main? I thought I fixed the issue with #7395

sohami commented:
> ```
> REPRODUCE WITH: ./gradlew ':server:internalClusterTest' --tests "org.opensearch.action.admin.cluster.stats.ClusterStatsIT.testNodeRolesWithDataNodeLegacySettings" -Dtests.seed=9C131CE66EBDF984 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=ar-TN -Dtests.timezone=Africa/Kinshasa -Druntime.java=19
> 
> org.opensearch.action.admin.cluster.stats.ClusterStatsIT > testNodeRolesWithDataNodeLegacySettings FAILED
>     java.lang.IllegalArgumentException: duplicate element: [cluster_manager]
>         at __randomizedtesting.SeedInfo.seed([9C131CE66EBDF984:B56ED209C37D3942]:0)
>         at java.****/java.util.ImmutableCollections$Set12.<init>(ImmutableCollections.java:791)
>         at java.****/java.util.Set.of(Set.java:487)
>         at org.opensearch.action.admin.cluster.stats.ClusterStatsIT.testNodeRolesWithDataNodeLegacySettings(ClusterStatsIT.java:415)
> ```
> 
> Strange! Seems like #7327 again. @sohami Did you rebase with the latest main? I thought I fixed the issue with #7395

I rebased 2 hours back and these commits were made an hour back so its not there. Rebased again

kotwanikunal commented:
Backport PR: https://github.com/opensearch-project/OpenSearch/pull/7408
